File: 0001-1000-easy-python/0225-implement-stack-using-queues-v2.py
# https://leetcode.com/problems/implement-stack-using-queues

from collections import deque
class MyStack(object):

    def __init__(self):
        self.queue = deque()
        self.front = None

    # O(1)
    def push(self, x: int) -> None:
        self.queue.appendleft(x)
        self.front = x

    # O(n)
    def pop(self) -> int:
        for i in range(len(self.queue) - 1):
            self.front = self.queue.pop()
            self.queue.appendleft(self.front)
        return self.queue.pop()

    # O(1)
    def top(self) -> int:
        return self.front

    # O(1)
    def empty(self) -> bool:
        return len(self.queue) == 0

# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: 0001-1000-easy-cpp/0225-implement-stack-using-queues-v2.cpp
// https://leetcode.com/problems/implement-stack-using-queues
class MyStack {
private:
    std::queue<int> queue;
    int front;

public:
    MyStack() {
    }

    // O(1)
    void push(int x) {
        this->queue.push(x);
        this->front = x;
    }

    // O(n)
    int pop() {
        for (int i = 0; i < this->queue.size() - 1; i++) {
            this->front = this->queue.front();
            this->queue.push(this->queue.front());
            this->queue.pop();
        }
        int x = this->queue.front();
        this->queue.pop();
        return x;
    }

    // O(1)
    int top() {
        return this->front;
    }

    // O(1)
    bool empty() {
        return this->queue.empty();
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */

